### PR TITLE
Helm Chart - support for priorityClassName

### LINKED ISSUE #20632: Add priorityClassName support
### Official Helm Chart version

1.2.0

### Apache Airflow version

2.1.4

### Kubernetes Version

1.21.2

### Helm Chart configuration

_No response_

### Docker Image customisations

_No response_

### What happened

_No response_

### What you expected to happen

_No response_

### How to reproduce

_No response_

### Anything else

It's currently impossible to assign a `priorityClassName` to the Airflow containers.
Seems like a very useful feature to ensure that the Airflow infrastructure has higher priority than "regular" pods.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by subkanthi**
closes: #20632 
Added support for priorityClassName in flower, webserver, pgbouncer, scheduler, statsd and triggerer and workers.
The user has to create the priority class outside of this chart and use the name in the priorityClassName fields.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: chart/values.yaml [hunk 1]
```diff
@@ -514,6 +514,7 @@ workers:
 
   # Select certain nodes for airflow worker pods.
   nodeSelector: {}
+  priorityClassName: ""
```
Comment (kaxil) [inline chart/values.yaml]:
```suggestion
  priorityClassName: ~
```

to be consistent, preference @jedcunningham ?

Comment (jedcunningham) [reply chart/values.yaml]: Yeah 👍

File: chart/templates/flower/flower-deployment.yaml [hunk 2]
```diff
@@ -67,6 +67,7 @@ spec:
       tolerations:
 {{ toYaml $tolerations | indent 8 }}
       serviceAccountName: {{ include "flower.serviceAccountName" . }}
+      priorityClassName: {{ .Values.flower.priorityClassName }}
```
Comment (jedcunningham) [inline chart/templates/flower/flower-deployment.yaml:71]: I think we should conditionally add these:

```suggestion
      {{- if .Values.flower.priorityClassName }}
      priorityClassName: {{ .Values.flower.priorityClassName }}
      {{- end }}
```

**Review by jedcunningham — CHANGES_REQUESTED** [review]
We also need tests added, maybe in `chart/tests/test_airflow_common.py` makes sense? Or we could do a separate test in each component's test file too.

**Comment by potiuk** [comment]
@subkanthi ?

**Comment by subkanthi** [comment]
> @subkanthi ?

Thanks Jarek, will get to it in a day or two.

File: chart/values.yaml (hunk 1 above)
Comment (subkanthi) [reply chart/values.yaml]: @jedcunningham , if I have it as tilde(~), then the validation fails, thoughts

```
Processing file:  chart/values.yaml
None is not of type 'string'

Failed validating 'type' in schema['properties']['workers']['properties']['priorityClassName']:
    {'default': '',
     'description': 'Set priority value from the priority class name',
     'type': 'string'}

On instance['workers']['priorityClassName']:
    None
None is not equal to the default of 

Failed validating 'default' in schema['properties']['workers']['properties']['priorityClassName']:
    {'default': '',
     'description': 'Set priority value from the priority class name',
     'type': 'string'}

On instance['workers']['priorityClassName']:
    None
```

File: chart/templates/flower/flower-deployment.yaml (hunk 2 above)
Comment (subkanthi) [reply chart/templates/flower/flower-deployment.yaml:71]: Changed.

File: chart/values.yaml (hunk 1 above)
Comment (jedcunningham) [reply chart/values.yaml]: You just need to change the type in the schema, e.g: https://github.com/apache/airflow/blob/0ac3b8c3dd749c59e60cf0169580b9e7c5049d9e/chart/values.schema.json#L431-L434

File: chart/templates/pgbouncer/pgbouncer-deployment.yaml [hunk 3]
```diff
@@ -59,6 +59,7 @@ spec:
         checksum/pgbouncer-config-secret: {{ include (print $.Template.BasePath "/secrets/pgbouncer-config-secret.yaml") . | sha256sum }}
         checksum/pgbouncer-certificates-secret: {{ include (print $.Template.BasePath "/secrets/pgbouncer-certificates-secret.yaml") . | sha256sum }}
     spec:
+      priorityClassName: {{ .Values.pgbouncer.priorityClassName }}
```
Comment (jedcunningham) [inline chart/templates/pgbouncer/pgbouncer-deployment.yaml:63]: I meant all instances of it should be conditional, not just flower :)

Comment (subkanthi) [reply chart/templates/pgbouncer/pgbouncer-deployment.yaml:63]: sure,  will make those changes.

File: chart/tests/test_airflow_common.py [hunk 4]
```diff
@@ -268,3 +268,46 @@ def test_have_all_config_mounts_on_init_containers(self):
...
+            elif component == "statsd":
+                assert priority == "low-priority-statsd"
+            elif component == "triggerer":
+                assert priority == "low-priority-triggerer"
+            elif component == "webserver":
+                assert priority == "low-priority-webserver"
+            elif component == "worker":
+                assert priority == "low-priority-worker"
+            else:
+                assert False
```
Comment (jedcunningham) [inline chart/tests/test_airflow_common.py]: Maybe this instead?
```suggestion
            assert priority == f"low-priority-{component}"
```

File: chart/values.schema.json [hunk 5]
```diff
@@ -1351,6 +1351,14 @@
                         "type": "string"
                     }
                 },
+                "priorityClassName": {
+                    "description": "Set priority value from the priority class name",
```
Comment (jedcunningham) [inline chart/values.schema.json]:
```suggestion
                    "description": "Specify priority for worker pods.",
```

This reads better to me?

File: chart/values.yaml [hunk 6]
```diff
@@ -638,6 +639,10 @@ scheduler:
   #      weight: 100
   tolerations: []
 
+  # Set the priority class name
+  # https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/#pod-priority
```
Comment (jedcunningham) [inline chart/values.yaml]:
```suggestion
```

I don't think we need these.

File: chart/tests/test_airflow_common.py (hunk 4 above)
Comment (subkanthi) [reply chart/tests/test_airflow_common.py]: Thanks, thought about this, but didnt realize it also covers the 'assert False condition.

File: chart/values.schema.json (hunk 5 above)
Comment (jedcunningham) [reply chart/values.schema.json]: I just suggested it once, but this applies to all of them too.

File: chart/values.yaml (hunk 6 above)
Comment (jedcunningham) [reply chart/values.yaml]: Same here.

Comment (subkanthi) [reply chart/values.yaml]: Removed.

File: chart/values.schema.json (hunk 5 above)
Comment (subkanthi) [reply chart/values.schema.json]: Updated all of them.

**Review by jedcunningham — APPROVED** [review]

**Pull request merged**